include/nlohmann/detail/input/binary_reader.hpp

```diff
@@ -120,6 +124,139 @@ class binary_reader
     }
 
   private:
+
+    /*!
+    @return whether array creation completed
+    */
+    bool get_bson_cstr(string_t& result)
+    {
+        auto out = std::back_inserter(result);
+        while (true)
+        {
+            get();
+            if (JSON_UNLIKELY(not unexpect_eof()))
+            {
+                return false;
+            }
+            if (current == 0x00)
+            {
+                return true;
+            }
+            *out++ = static_cast<char>(current);
+        }
+
+        return true;
+    }
+
+    bool parse_bson_entries(bool is_array)
+    {
+        while (auto entry_type = get())
+        {
+            string_t key;
+            if (!get_bson_cstr(key))
+            {
+                return false;
+            }
+
+            if (!is_array)
+            {
+                sax->key(key);
+            }
+
+            switch (entry_type)
+            {
+                case 0x01: // double
+                {
+                    double number;
+                    get_number<double, true>(number);
+                    sax->number_float(static_cast<number_float_t>(number), "");
+                }
+                break;
+                case 0x02: // string
+                {
+                    std::int32_t len;
+                    string_t value;
+                    get_number<std::int32_t, true>(len);
+                    get_string(len - 1ul, value);
+                    get();
+                    sax->string(value);
+                }
+                break;
+                case 0x08: // boolean
+                {
+                    sax->boolean(static_cast<bool>(get()));
+                }
+                break;
+                case 0x10: // int32
+                {
+                    std::int32_t value;
+                    get_number<std::int32_t, true>(value);
+                    sax->number_integer(static_cast<std::int32_t>(value));
+                }
+                break;
+                case 0x12: // int64
+                {
+                    std::int64_t value;
+                    get_number<std::int64_t, true>(value);
+                    sax->number_integer(static_cast<std::int64_t>(value));
+                }
+                break;
+                case 0x0A: // null
+                {
+                    sax->null();
+                }
+                break;
+                case 0x03: // object
+                {
+                    parse_bson_internal();
+                }
+                break;
+                case 0x04: // array
+                {
+                    parse_bson_array();
+                }
+                break;
+            }
```

Comment:
What about all the other values for `entry_type`?

Reply:
added a `default:` clause which will raise a new parse_error.114: Unsupported BSON record type. Also improved the error handling for the other cases, where in most occasions, the return values were ignored before.